File: python/src/wdl_model/model/base/wdl_key_value.py
"""Generic key-value model nodes used across the Python WDL object model."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Generic, Optional, TypeVar

from .wdl_node import WdlNode

K = TypeVar("K")
V = TypeVar("V")


@dataclass
class WdlKeyValue(Generic[K, V], WdlNode):
    """Generic key-value node.

    Examples include metadata entries, runtime and requirements entries, call input bindings, and
    map/object-like literal entries.
    """

    key: Optional[K] = None
    value: Optional[V] = None

    def getKey(self) -> Optional[K]:
        """Return the key portion of the pair."""
        return self.key

    def setKey(self, key: Optional[K]) -> None:
        """Set the key portion of the pair."""
        self.key = key

    def getValue(self) -> Optional[V]:
        """Return the value portion of the pair."""
        return self.value

    def setValue(self, value: Optional[V]) -> None:
        """Set the value portion of the pair."""
        self.value = value


class WdlStringKeyValue(WdlKeyValue[str, "WdlExpression"]):
    """Key-value node whose key is a string and whose value is an expression."""

    pass


class WdlExpresionKeyValue(WdlKeyValue["WdlExpression", "WdlExpression"]):
    """Key-value node whose key and value are both expressions."""

    pass

File: java/src/main/java/com/myriad/wdl/model/base/WdlKeyValue.java
package com.myriad.wdl.model.base;

import com.myriad.wdl.model.expressions.WdlExpression;
import lombok.Getter;
import lombok.Setter;

/**
 * Generic key-value node used throughout the WDL model.
 *
 * <p>Examples include metadata entries, runtime and requirements entries, map literal entries, and
 * call input bindings. The WDL specification uses key-value structures in sections such as
 * metadata, requirements, hints, and JSON-like literal forms.
 */
public abstract class WdlKeyValue<K, V> implements WdlNode {
  @Getter @Setter private K key;
  @Getter @Setter private V value;

  protected WdlKeyValue() {}

  protected WdlKeyValue(K key) {
    this();
    setKey(key);
  }

  protected WdlKeyValue(K key, V value) {
    this(key);
    setValue(value);
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }

  public abstract static class WdlStringKeyValue extends WdlKeyValue<String, WdlExpression> {
    protected WdlStringKeyValue() {
      super();
    }

    protected WdlStringKeyValue(String key, WdlExpression value) {
      super(key, value);
    }

    protected WdlStringKeyValue(String key) {
      super(key);
    }
  }

  /** Key-value node whose key and value are both expressions. */
  public abstract static class WdlExpresionKeyValue
      extends WdlKeyValue<WdlExpression, WdlExpression> {
    protected WdlExpresionKeyValue() {
      super();
    }

    protected WdlExpresionKeyValue(WdlExpression key, WdlExpression value) {
      super(key, value);
    }

    protected WdlExpresionKeyValue(WdlExpression key) {
      super(key);
    }
  }
}
